File: ahoy-ui/src/app/applications/application.ts
/*
 * Copyright  2021 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import {EnvironmentReleaseId} from '../environment-release/environment-release';
import {ReleaseVersion} from '../releases/release';

export class Application {
  id: number;
  name: string;
  applicationVersions: ApplicationVersion[];
  latestApplicationVersion: ApplicationVersion;
}

export class ApplicationVersion {
  id: number;
  version: string;
  spec: ApplicationSpec;
  application: Application | string;
  status: ApplicationReleaseStatus;
  releaseVersions: ReleaseVersion[];
}

export class ApplicationSpec {
  image: string;
  dockerRegistryName: string;
  servicePorts: number[] = [];
  healthEndpointPath: string;
  healthEndpointPort: number;
  healthEndpointScheme = 'HTTP';
  environmentVariables: ApplicationEnvironmentVariable[] = [];
  configPath: string;
  configFiles: ApplicationConfigFile[] = [];
  volumes: ApplicationVolume[] = [];
  secrets: ApplicationSecret[] = [];
}

export class ApplicationEnvironmentVariable {
  key: string;
  value: string;
  type: string;
  secretName: string;
  secretKey: string;

  static newValueType(key: string, value: string): ApplicationEnvironmentVariable {
    const environmentVariable = new ApplicationEnvironmentVariable();
    environmentVariable.type = 'Value';
    environmentVariable.key = key;
    environmentVariable.value = value;
    return environmentVariable;
  }

  static newSecretType(key: string, secretName: string, secretKey: string): ApplicationEnvironmentVariable {
    const environmentVariable = new ApplicationEnvironmentVariable();
    environmentVariable.type = 'Secret';
    environmentVariable.key = key;
    environmentVariable.secretName = secretName;
    environmentVariable.secretKey = secretKey;
    return environmentVariable;
  }
}

export class ApplicationConfigFile {
  name: string;
  content: string;
}

export class ApplicationVolume {
  name: string;
  mountPath: string;
  type: string;
  storageClassName = 'standard';
  accessMode = 'ReadWriteOnce';
  size: number;
  sizeStorageUnit = 'Gi';
  secretName: string;
}

export class ApplicationSecret {
  name: string;
  type: string;
  data: { [key: string]: string };
}

export class ApplicationEnvironmentConfig {
  id: ApplicationEnvironmentConfigId;
  spec: ApplicationEnvironmentSpec;
}

export class ApplicationEnvironmentSpec {
  replicas = 1;
  routeHostname: string;
  routeTargetPort: number;
  tls: boolean;
  tlsSecretName: string;
  environmentVariables: ApplicationEnvironmentVariable[] = [];
  configFiles: ApplicationConfigFile[] = [];
  volumes: ApplicationVolume[] = [];
  secrets: ApplicationSecret[] = [];
}

export class ApplicationEnvironmentConfigId {
  environmentReleaseId: EnvironmentReleaseId;
  releaseVersionId: number;
  applicationVersionId: number;
}

export class ApplicationEnvironmentConfigIdUtil {

  public static toIdString(applicationEnvironmentConfig: ApplicationEnvironmentConfig): string {
    const id = applicationEnvironmentConfig.id;
    const erId = id.environmentReleaseId;
    return `${erId.environmentId}_${erId.releaseId}_${id.releaseVersionId}_${id.applicationVersionId}`;
  }

  public static toIdStringFromId(id: ApplicationEnvironmentConfigId): string {
    const erId = id.environmentReleaseId;
    return `${erId.environmentId}_${erId.releaseId}_${id.releaseVersionId}_${id.applicationVersionId}`;
  }
}

export class ApplicationReleaseStatus {
  id: ApplicationEnvironmentConfigId;
  status: string;
}

File: ahoy-components/src/main/java/za/co/lsd/ahoy/server/applications/Application.java
/*
 * Copyright  2022 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package za.co.lsd.ahoy.server.applications;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import org.hibernate.Hibernate;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@Getter
@Setter
@ToString
@RequiredArgsConstructor
public class Application {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;
	@Column(unique = true)
	@NotNull
	@Pattern(regexp = "^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$",
		message = "Name invalid: should start with and use lower case letters and numbers")
	@Size(max = 53)
	private String name;

	@OneToMany(mappedBy = "application", cascade = CascadeType.REMOVE)
	@JsonIgnore
	@OrderBy("id")
	@ToString.Exclude
	private List<ApplicationVersion> applicationVersions = new ArrayList<>();

	public ApplicationVersion latestApplicationVersion() {
		if (applicationVersions != null && applicationVersions.size() > 0) {
			return applicationVersions.get(applicationVersions.size() - 1);
		}
		return null;
	}

	public Application(@NotNull String name) {
		this.name = name;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || Hibernate.getClass(this) != Hibernate.getClass(o)) return false;
		Application that = (Application) o;
		return Objects.equals(id, that.id);
	}

	@Override
	public int hashCode() {
		return Objects.hash(id);
	}
}
